ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/Genesis.hs

```diff
@@ -0,0 +1,589 @@
+      separateDeclinedAndStillInRace ::
+        [(FetchDecision (ChainSuffix header), peerInfo)] ->
+        WithDeclined peerInfo (Maybe (ChainSuffix header, [(ChainSuffix header, peerInfo)]))
+      separateDeclinedAndStillInRace decisions = do
+        let (declined, inRace) = partitionEithers
+              [ bimap ((,p)) ((,p)) d | (d, p) <- decisions ]
+        tell (DList.fromList declined)
+        case inRace of
+          [] -> pure Nothing
+          _ : _ -> do
+            let maxChainOn f c0 c1 = case compareCandidateChains (f c0) (f c1) of
+                  LT -> c1
+                  _  -> c0
+                -- maximumBy yields the last element in case of a tie while we
+                -- prefer the first one
+                chainSfx = fst $
+                  List.foldl1' (maxChainOn (getChainSuffix . fst)) inRace
+            pure $ Just (chainSfx, inRace)
+
+-- | Given _the_ candidate fragment to sync from, and a list of peers (with
+-- their corresponding candidate fragments), choose which peer to sync _the_
+-- candidate fragment from.
+--
+-- We first filter out all the peers that cannot even serve a reasonable batch
+-- of _the_ candidate fragment, and then we choose the first one according to
+-- the ordering passed as argument.
+--
+-- PRECONDITION: The set of peers must be included in the peer order queue.
+-- PRECONDITION: The given candidate fragments must not be empty.
+selectThePeer ::
+  forall header peer extra.
+  HasHeader header =>
+  -- | The candidate fragment that we have selected to sync from, as suffix of
+  -- the immutable tip.
+  CandidateFragments header ->
+  -- | Association list of candidate fragments (as suffixes of the immutable
+  -- tip) and their associated peers.
+  [(ChainSuffix header, PeerInfo header peer extra)] ->
+  WithDeclined
+    (PeerInfo header peer extra)
+    (Maybe (ChainSuffix header, PeerInfo header peer extra))
+selectThePeer
+  theFragments
+  candidates = do
+    -- Create a fetch request for the blocks in question. The request has exactly
+    -- 1 block. It will only be used to choose the peer to fetch from, but we will
+    -- later craft a more refined request for that peer. See [About the gross
+    -- request] in the module documentation. Because @theFragments@ is not
+    -- empty, and does not contain empty fragments, @grossRequest@ will not be empty.
+    let firstBlock = map (AF.takeOldest 1) . take 1 . filter (not . AF.null)
+        requestBlock = firstBlock $ snd theFragments
+        grossRequest = FetchRequest $ assert (all (not . AF.null) requestBlock) requestBlock
+
+    -- Return the first peer that can serve the gross request and decline
+    -- the other peers.
+    go grossRequest candidates
+  where
+    go grossRequest (c@(candidate, peerInfo) : xs) = do
+      if requestHeadInCandidate candidate grossRequest then do
+        tell $ DList.fromList
+          [(FetchDeclineConcurrencyLimit FetchModeBulkSync 1, pInfo)
```

Comment:
Should we use `FetchModeGenesis` instead? This would require some changes in `Decision.hs` to accomodate for the type `GenesisFetchMode`.

Reply:
It indeed seems nicer to use `FetchModeGenesis` here, I think very small changes to `Decision.hs` (especially if they are in a separate commit) are fine.